src/Monitor.c

```diff
@@ -48,8 +53,72 @@ static bool handle_activation_log_cyclic_buffer(const enum interfaces_enum inter
     return true;
 }
 
+static bool cpu_usage_visitor(Thread_Control *the_thread, void *arg)
+{
+    float usage_percent;
+	uint32_t integer_val;
+	uint32_t float_val;
+    Timestamp_Control uptime;
+	const Timestamp_Control used_time =
+	    _Thread_Get_CPU_time_used_after_last_reset(the_thread);
+
+	_TOD_Get_uptime(&uptime);
+	_Timestamp_Subtract(&uptime_at_last_reset, &uptime, &total_usage_time);
+	_Timestamp_Divide(&used_time, &total_usage_time, &integer_val, &float_val);
+
+	usage_percent = (float)float_val / TOD_NANOSECONDS_PER_MICROSECOND;
+	usage_percent += (float)integer_val;
+
+	if (usage_percent < idle_cpu_usage_data.minimum_cpu_usage) {
+		idle_cpu_usage_data.minimum_cpu_usage = usage_percent;
+	}
+
+	if (usage_percent > idle_cpu_usage_data.maximum_cpu_usage) {
+		idle_cpu_usage_data.maximum_cpu_usage = usage_percent;
+	}
+
+	idle_cpu_usage_data.average_cpu_usage =
+	    idle_cpu_usage_data.average_cpu_usage +
+	    (usage_percent - idle_cpu_usage_data.average_cpu_usage) /
+		(benchmarking_ticks + 1);
+
+    // only first idle thread is needed, stop iteration after first step
+    return true;
+}
+
 bool Monitor_Init()
 {
+    _Timestamp_Set_to_zero(&total_usage_time);
+    rtems_cpu_usage_reset();
+	_TOD_Get_uptime(&uptime_at_last_reset);
+	
+    for(int i = 0; i < RUNTIME_THREAD_COUNT; i++){
+        idle_cpu_usage_data.maximum_cpu_usage = 0.0;
+        idle_cpu_usage_data.minimum_cpu_usage = FLT_MAX;
+        idle_cpu_usage_data.average_cpu_usage = 0.0;
+    }
+
+    return true;
+}
+
+bool Monitor_MonitoringTick(void)
+{
+	rtems_task_iterate(cpu_usage_visitor, NULL);
+    benchmarking_ticks++;
+}
+
+bool Monitor_GetUsageData(const enum interfaces_enum interface, struct Monitor_InterfaceUsageData *const usage_data)
+{
+    usage_data->interface = interface;
+    usage_data->maximum_execution_time = threads_info[interface].max_thread_execution_time;
```

Comment:
Where are these values updated?

Reply:
The values are calculated in runtime generated in kazoo. Here there was only need for data copy.